lib/xwax/pitch.h

```diff
@@ -1,71 +1,213 @@
+
 /*
- * Copyright (C) 2021 Mark Hills <mark@xwax.org>
- *
- * This file is part of "xwax".
- *
- * "xwax" is free software; you can redistribute it and/or modify it
- * under the terms of the GNU General Public License, version 3 as
- * published by the Free Software Foundation.
- *
- * "xwax" is distributed in the hope that it will be useful, but
- * WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
- * General Public License for more details.
+ * Kalman filter replacement for alpha-beta pitch estimator
+ * Model: constant velocity (x,v) with acceleration process noise
  *
- * You should have received a copy of the GNU General Public License
- * along with this program; if not, see <https://www.gnu.org/licenses/>.
+ * State:
+ *   x: displacement over the last interval (kept relative by subtracting dx at the end)
+ *   v: velocity (pitch)
  *
+ * Observation:
+ *   z = dx  (measured displacement over dt)
  */
 
 #ifndef PITCH_H
 #define PITCH_H
 
-/* Values for the filter concluded experimentally */
+#include <math.h>
+#include <errno.h>
+#include <stdio.h>
 
-#define ALPHA (1.0/512)
-#define BETA (ALPHA/256)
+struct pitch {
 
-/* State of the pitch calculation filter */
+    /* configuration */
+    double dt;     /* sample period (s) */
+    double *q;      /* process noise spectral density (acceleration), e.g. (rad/s^2)^2 * s */
+    double *r;      /* measurement noise variance for dx (units^2) */
 
-struct pitch {
-    double dt, x, v;
+    /* state */
+    double x;      /* position (relative; we subtract dx each step) */
```

Comment:
Can you also document the physical unit for all values below.

Reply:
There is not really a physical unit here. The velocity is the pitch ranging from being a signed double, where 0.0 is no movement, 1.0 steady playback and -1.0 reverse playback. The position is the current value of the carrier wave. Units don't really help here to understand the filter.

Reply:
> The velocity is the pitch ranging from being a signed double, where 0.0 is no movement, 1.0 steady playback and -1.0 reverse playback.

This sounds like what we call "rate" at other places of the Mixxx code. Does XWAX uses the term "velocity" for this?

> The position is the current value of the carrier wave.

This explanation is not clear to me, is the position is in 2*Pi or in samples, or the amplitude of the sine wave at the current position?

Reply:
If XWax uses this notation, than you should keep it!